lib/proxy/conn_test.go

```diff
@@ -0,0 +1,204 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy
+
+import (
+	"context"
+	"net"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/stretchr/testify/require"
+)
+
+type mockStream struct {
+	ctx  context.Context
+	conn net.Conn
+}
+
+func newMockStream(ctx context.Context, conn net.Conn) *mockStream {
+	return &mockStream{
+		ctx:  ctx,
+		conn: conn,
+	}
+}
+
+func (m *mockStream) Context() context.Context {
+	return m.ctx
+}
+
+func (m *mockStream) Send(f *proto.Frame) error {
+	_, err := m.conn.Write(f.GetData().Bytes)
+	return err
+}
+
+func (m *mockStream) Recv() (*proto.Frame, error) {
+	b := make([]byte, 2*maxChunkSize)
+	n, err := m.conn.Read(b)
+	return &proto.Frame{Message: &proto.Frame_Data{Data: &proto.Data{
+		Bytes: b[:n],
+	}}}, err
+}
+
+func newStreamPipe() (*streamConn, net.Conn) {
+	local, remote := net.Pipe()
+	stream := newMockStream(context.Background(), remote)
+
+	timeout := time.Now().Add(time.Second * 5)
+	local.SetReadDeadline(timeout)
+	local.SetWriteDeadline(timeout)
+	remote.SetReadDeadline(timeout)
+	remote.SetWriteDeadline(timeout)
+
+	src := &utils.NetAddr{Addr: "src"}
+	dst := &utils.NetAddr{Addr: "dst"}
+	streamConn := newStreamConn(stream, src, dst)
+
+	return streamConn, local
+}
+
+func TestStreamConnWrite(t *testing.T) {
+	streamConn, local := newStreamPipe()
+	wg := &sync.WaitGroup{}
+	wg.Add(2)
+
+	data := []byte("hello world!")
+	go func() {
+		defer wg.Done()
+		n, err := streamConn.Write(data)
+		require.NoError(t, err)
+		require.Equal(t, len(data), n)
```

Comment:
Assertions in spawned goroutines will cause problems due to semantics of `t.FailNow`. Here and in all other occurrences in this file the data needs to be sent over a channel and asserted on the main goroutine that is running the test

Reply:
So I believe using assert fixes this since it calls `testing.Errorf` and `testing.Log` under the hood which allows the test to continue and fails at the end. Updated here [7c047ff](https://github.com/gravitational/teleport/pull/10223/commits/7c047ffcb36cfb231334a68000fa7a9cb38cd7a9). If its preferred to use `require` I can move all the checks to the the main goroutine but using assert feels a little easier to read/reason about in this case.